include/measurement_kit/swig.hpp

```diff
@@ -0,0 +1,167 @@
+// Part of Measurement Kit <https://measurement-kit.github.io/>.
+// Measurement Kit is free software under the BSD license. See AUTHORS
+// and LICENSE for more information on the copying conditions.
+#ifndef MEASUREMENT_KIT_SWIG_HPP
+#define MEASUREMENT_KIT_SWIG_HPP
+
+/*
+ * Measurement Kit C++ interface for SWIG.
+ *
+ * See include/measurement_kit/engine.h for more documentation.
+ *
+ * See example/swig/ndt.cpp for example.
+ */
+
+#include <measurement_kit/ffi.h>
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+namespace mk {
+namespace swig {
+
+/*
+    Design/implementation/integration notes:
+
+    1. To use this class with SWIG, `%include "std_string.i"`. We cannot avoid
+       using `std::string`, because we need to express the concept that the
+       caller takes ownership of the value returned by wait_for_next_event().
+
+    2. Move semantic is not wrapped by SWIG. We replace it with passing a
+       constant reference for the settings of the Task.
+
+    3. Parsing of the stringified JSON can throw. We have therefore created a
+       specific method for initializing, given that throwing in general is
+       something that makes life more complicate for bindings. (We may still
+       fail and abort() in case of bad error conditions, but a parse error
+       is clearly an avoidable error.)
+
+    4. This class is implemented in terms of the FFI API by design choice, so
+       in specific cases MK can export just the ANSI-C API.
+
+    Example minimal SWIG interface file:
+
+    ```
+    %module measurement_kit
+    %include "std_string.i"
+    %{
+    #include <measurement_kit/swig.hpp>
+    %}
+    %include <measurement_kit/swig.hpp>
+    ```
+*/
+
+/// InitializeExResult is the result returned by Task::initialize_ex().
+class InitializeExResult {
+  public:
+    /// reason is the error that occurred.
+    std::string reason;
+    /// result tells you whether the call succeeded.
+    bool result = false;
+};
+
+/// Task is something that Measurement Kit can do.
+class Task {
+  public:
+
+// Prevent SWIG from attempting to generate wrappers for these.
+#ifndef SWIG
+    class TaskDeleter {
+      public:
+        void operator()(mk_task_t *p) {
+            if (p != nullptr) {
+                mk_task_destroy(p);
+            }
+        }
+    };
+    class EventDeleter {
+      public:
+        void operator()(mk_event_t *p) {
+            if (p != nullptr) {
+                mk_event_destroy(p);
+            }
+        }
+    };
+#endif
+
+    /// Task() creates an empty task.
+    Task() {}
+
+    /// initialize() initializes a Task with @p settings. @param settings is
```

Comment:
Thanks, it didn't occur to me that this could happen! So, I just checked with Doxygen and it seems that it is actually able to handle this case:

![screen shot 2018-02-27 at 12 08 07](https://user-images.githubusercontent.com/337298/36725580-02f92270-1bb7-11e8-84d5-74f7330eee5b.png)